src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
+            throw new IndexOutOfBoundsException("Wrong index");
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(last, value, null);
+        Node<T> oldNode = last;
+        last = node;
+        if (size == 0) {
+            first = node;
+        } else {
+            oldNode.next = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size == 0 && index == 0) {
+            Node<T> node = new Node<>(null, value, null);
+            first = last = node;
+            size++;
+            return;
+        }
+        if (index == 0) {
+            Node<T> node = new Node<>(null, value, first);
+            Node<T> oldNode = first;
+            first = oldNode.prev = node;
+            size++;
+            return;
+        }
+        if (index == size) {
+            Node<T> node = new Node<>(last, value, null);
+            Node<T> oldNode = last;
+            last = oldNode.next = node;
+            size++;
+            return;
+        }
+        checkOutIndex(index);
+        Node<T> oldNode = first;
+        int i = 0;
+        while (i < index) {
+            oldNode = oldNode.next;
+            i++;
+        }
+        Node<T> node = new Node<>(oldNode.prev, value, oldNode);
+        oldNode.prev.next = oldNode.prev = node;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int i = 0;
+        while (i < list.size()) {
+            add(list.get(i));
+            i++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkOutIndex(index);
+        Node<T> getNode = first;
+        int i = 0;
+        while (i < index) {
+            getNode = getNode.next;
+            i++;
+        }
+        return getNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkOutIndex(index);
+        Node<T> setNode = first;
+        int i = 0;
+        while (i < index) {
+            setNode = setNode.next;
+            i++;
+        }
+        T oldValue = setNode.value;
+        setNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkOutIndex(index);
+        if (index == 0) {
+            final T removeValue = first.value;
+            if (size == 1) {
+                first = last = null;
+            } else {
+                first = first.next;
+                first.prev = null;
+            }
+            size--;
+            return removeValue;
+        }
+        if (index == size - 1) {
+            final T removeValue = last.value;
+            last = last.prev;
+            last.next = null;
+            size--;
+            return removeValue;
+        }
+        int i = 0;
+        Node<T> removeNode = first;
+        while (i < index) {
+            removeNode = removeNode.next;
+            i++;
```

Comment:
the logic of this method is too complex, let's simplify it. This method should make two things:
1. Find deleted node by index
2. Unlink node from its neighbors and link these neighbors with each other